### PR TITLE
Feat/request payment modal 16736

### LINKED ISSUE #16736: Request payment - modal
Implement modal for Request Payment

### PR DESCRIPTION
**Pull request by Cuteivist**
Closes #16736

### What does the PR do

* Added feature flag for Request Payment. Disabled until whole feature is finished
* Added Request payment modal
* Added menu for message commands (now featuring add image and add request payment). Request payment is disabled for testnet to prevent corner cases or issues.
* Added RequestPaymentStore to easily pass required models and stores to modal
* Updated storybook

### Architecture compliance

- [X] I am familiar with the application architecture and agreed good practices.
My PR is consistent with this document: [Status Desktop Architecture Guide](https://github.com/status-im/status-desktop/blob/master/CONTRIBUTING.md)

### Screenshot of functionality (including design for comparison)

- [X] I've checked the design and this PR matches it

image

image

### Impact on end user

What is the impact of these changes on the end user (before/after behaviour)

### How to test

- How should one proceed with testing this PR.
Enable request payment feature flag, go to any chat on mainnet.
- What kind of user flows should be checked?

### Risk 

Described potential risks and worst case scenarios.

Tick **one**:
- [X] Low risk: 2 devs MUST perform testing as specified above and attach their results as comments to this PR **before** merging.
- [ ] High risk: QA team MUST perform additional testing in the specified affected areas **before** merging.

### CODE REVIEW

File: ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        AmountToSend {
+            id: amountToSendInput
+            Layout.fillWidth: true
+
+            readonly property bool ready: valid && !empty
+            readonly property double asNumber: {
+                if (!valid)
+                    return 0
+
+                return parseFloat(text.replace(LocaleUtils.userInputLocale.decimalPoint, "."))
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml:112]: Are we fine with loss of precision here? :thinking:

**Review by caybro — COMMENTED** [review]
LGTM in general, apart from the vertical (mis)alignment between the amount and the token selector when looking at the screenshot :)

**Comment by alexjba** [comment]
Build fails on all platforms 
```
[2024-11-11T14:30:45.449Z] /home/jenkins/workspace/rs_linux_x86_64_package_PR-16744/src/app/global/feature_flags.nim(13, 1) template/generic instantiation of `QtObject` from here
[2024-11-11T14:30:45.450Z] /home/jenkins/workspace/rs_linux_x86_64_package_PR-16744/src/app/global/feature_flags.nim(26, 32) Error: undeclared field: 'requestPaymentEnabled=' for type feature_flags.FeatureFlags [type declared in /home/jenkins/workspace/rs_linux_x86_64_package_PR-16744/src/app/global/feature_flags.nim(14, 8)]
```

File: storybook/pages/RequestPaymentModalPage.qml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        id: popupBg
+
+        SplitView.fillWidth: true
+        SplitView.fillHeight: true
+
+        Button {
+            id: reopenButton
+            anchors.centerIn: parent
+            text: "Reopen"
+            enabled: !requestPaymentModalComponent.visible
```
Comment (alexjba) [inline storybook/pages/RequestPaymentModalPage.qml:73]: Probably not needed. Also, here we're using the `Component` and `visible` is undefined.

File: ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+import shared.popups.send.views 1.0
+import shared.controls 1.0
+import shared.stores 1.0 as SharedStores
+import utils 1.0
+
+StatusDialog {
+    id: root
+
+    required property SharedStores.RequestPaymentStore store
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml:24]: WDYT of removing the store dependency here and replace it with properties? Would be much clearer what is needed for this Modal.

I see we're not using many properties/functions from the store and could be easely done.

File: ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        id: d
+
+        readonly property ModelEntry selectedHolding: ModelEntry {
+            sourceModel: holdingSelector.model
+            key: "tokensKey"
+        }
+
+        readonly property bool isSelectedHoldingValidAsset: !!selectedHolding.item
+
+        readonly property var adaptor: TokenSelectorViewAdaptor {
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml:67]: We have this warning in storybook:
`file:///status-desktop/ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:285:17: Unable to assign [undefined] to QAbstractItemModel*`

File: ui/imports/shared/status/StatusChatInput.qml [hunk 5]
```diff
@@ -985,13 +1005,45 @@ Rectangle {
...
+        }
+    }
+
+    Component {
+        id: requestPaymentPopupComponent
+        RequestPaymentModal {
+            store: control.requestPaymentStore
+
+            onAccepted: {
+                control.requestPaymentStore.addPaymentRequest(selectedTokenKey, amount, selectedAccountAddress, selectedNetworkChainId)
```
Comment (alexjba) [inline ui/imports/shared/status/StatusChatInput.qml:1023]: Missing `requestPaymentStore.addPaymentRequest` implementation.

File: ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml
Comment (alexjba) [inline ui/app/AppLayouts/Chat/popups/RequestPaymentModal.qml:1]: Please add qml tests for the new component (or add a task to do it before we need to add more features on top of this)

**Review by alexjba — COMMENTED** [review]
Nice work! LGTM in general!

A few findings below.

**Review by micieslak — DISMISSED** [review]
Definitely good job in general, however I have several concerns regarding the architecture which are worth amending imo:

1. The modal should be fully decoupled from `StatusChatInput`. `StatusChatInput` should only emit signal with request to launch the modal instead of managing it internally, because it causes that `StatusChagInput` depends on everything the modal depends on now and will depend on in the future. Those two things become tightly coupled unnecessarily.
2. `RequestPaymentStore` is probably not needed at all. Broadly speaking, the store is not intended to be a container to move stuff around UI components. It's to expose backend's functionality. And every backend's mode/signal/function should be exposed only once via a single store.
3. The `RequestPaymentModal` should be a plain UI component taking some models and emitting some signals with no dependency on any store. Also transformations like that done by `TokenSelectorViewAdaptor` should not be done internally. Instead the modal should depend on the model. It's up to the caller to compose data using any adaptor that's needed.

In case of any doubts please ping me, I will be happy to discuss about that.

File: src/app/global/feature_flags.nim [hunk 6]
```diff
@@ -5,7 +5,7 @@ const DEFAULT_FLAG_DAPPS_ENABLED = true
 const DEFAULT_FLAG_SWAP_ENABLED = true
 const DEFAULT_FLAG_CONNECTOR_ENABLED* = false
 const DEFAULT_FLAG_SEND_VIA_PERSONAL_CHAT_ENABLED = true
-const DEFAULT_FLAG_PAYMENT_REQUEST_ENABLED = true
+const DEFAULT_FLAG_PAYMENT_REQUEST_ENABLED = false
```
Comment (caybro) [inline src/app/global/feature_flags.nim:8]: Intended?

File: ui/imports/utils/Global.qml [hunk 7]
```diff
@@ -92,7 +92,7 @@ QtObject {
 
     signal privacyPolicyRequested()
 
-    signal openPaymentRequestModalRequested()
+    signal openPaymentRequestModalRequested(var inputAreaModule)
```
Comment (caybro) [inline ui/imports/utils/Global.qml:95]: Ugh, that's nasty... passing the NIM backend module around with a signal :)

File: ui/app/mainui/Popups.qml [hunk 8]
```diff
@@ -53,7 +53,6 @@ QtObject {
     property WalletStores.CollectiblesStore walletCollectiblesStore
     property NetworkConnectionStore networkConnectionStore
     property WalletStores.BuyCryptoStore buyCryptoStore
-    property ChatStores.RootStore chatStore
```
Comment (caybro) [inline ui/app/mainui/Popups.qml]: This definitely must stay!

**Review by caybro — CHANGES_REQUESTED** [review]
LGTM in general, besides the 2 things in Popups.qml

File: ui/imports/utils/Global.qml (hunk 7 above)
Comment (alexjba) [reply ui/imports/utils/Global.qml:95]: The alternative now would be to register a callback in the signal as we have for other popups. Then we can connect to `onAccepted`.

File: ui/app/AppLayouts/Chat/popups/PaymentRequestModal.qml [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                value: root.selectedNetworkChainId ?? -1
+            }
+
+            model: root.flatNetworksModel
+
+            control.background: SQP.StatusComboboxBackground {
+                active: networkSelector.control.down || networkSelector.control.hovered
+            }
+
+            component NetworkDelegate: StatusListItem {
```
[the remark this answers is not in the document]
Comment (alexjba) [reply ui/app/AppLayouts/Chat/popups/PaymentRequestModal.qml:203]: The tests are probably failing for the same reason. On CI it runs on qt 5.15.2

[the remark this answers is not in the document]
Comment (alexjba) [reply ui/app/AppLayouts/Chat/popups/PaymentRequestModal.qml]: It's quite the effort, but hopefully we'll have it in 2.33.

**Review by alexjba — APPROVED** [review]
LGTM! 👍

**Review by caybro — APPROVED** [review]
LGTM!

File: ui/imports/shared/popups/send/views/AmountToSend.qml [hunk 10]
```diff
@@ -86,6 +87,9 @@ Control {
     property var formatBalance: balance =>
                                 `${balance.toLocaleString(Qt.locale())} CRYPTO`
 
+    /* Shows separator between top and bottom text */
+    property bool showSeparator: false
```
Comment (micieslak) [inline ui/imports/shared/popups/send/views/AmountToSend.qml:91]: There is already one similar flag `dividerVisible`. Please user similar name and group them close to each-other. And please add switch in SB page.

Comment (micieslak) [reply ui/imports/shared/popups/send/views/AmountToSend.qml:91]: Or it's even the same thing and should be de-duplicated?

File: ui/imports/shared/status/StatusChatInput.qml [hunk 11]
```diff
@@ -67,6 +72,7 @@ Rectangle {
     property var fileUrlsAndSources: []
 
     property var linkPreviewModel: null
+    property var paymentRequestModel: null
```
Comment (micieslak) [inline ui/imports/shared/status/StatusChatInput.qml:75]: is it used?

File: ui/imports/shared/status/StatusChatInput.qml [hunk 12]
```diff
@@ -985,13 +998,45 @@ Rectangle {
...
+                text: qsTr("Add image")
+                icon.name: "image"
+                onTriggered: control.openImageDialog()
+            }
+
+            MouseArea {
+                implicitWidth: paymentRequestMenuItem.width
+                implicitHeight: paymentRequestMenuItem.height
+                hoverEnabled: true
+                visible: control.paymentRequestEnabled
```
Comment (micieslak) [inline ui/imports/shared/status/StatusChatInput.qml:1025]: The naming is a bit confusing here. With `paymentRequestEnabled` set to true we can have payment actually disabled.

Maybe pair like:
```
bool paymentRequestsVisible
bool paymentRequestsDisabledInTestMode
```
(probably still not ideal but maybe bit better)

File: ui/app/mainui/AppMain.qml [hunk 13]
```diff
@@ -1764,6 +1768,7 @@ Item {
                                 transactionStore: appMain.transactionStore
                                 walletAssetsStore: appMain.walletAssetsStore
                                 currencyStore: appMain.currencyStore
+                                paymentRequestEnabled: featureFlagsStore.paymentRequestEnabled
```
Comment (micieslak) [inline ui/app/mainui/AppMain.qml:1771]: Are payment requests to be available for both personal chats and communities?

File: ui/imports/utils/Global.qml [hunk 14]
```diff
@@ -92,6 +92,8 @@ QtObject {
 
     signal privacyPolicyRequested()
 
+    signal openPaymentRequestModalRequested(var callback)
```
Comment (micieslak) [inline ui/imports/utils/Global.qml:95]: The preferred way is, in general, explicit signals propagation (like in https://github.com/status-im/status-desktop/pull/16718) but I think it's ok for now.

**Review by micieslak — COMMENTED** [review]
Hey! Thank you for all those alignments you introduced. Looks much better. It's ok in general, just found some minor things that should be quick to fix.

**Review by virginiabalducci — APPROVED** [review]
This looks good!
Testing feedback here: https://github.com/status-im/status-desktop/pull/16833
The pending issues are not related to this feature.

**Pull request merged**